### PR TITLE
osd: make osd recover more smoothly by avoiding failure peer info to resent

### PR DESCRIPTION
**Pull request by shun-s**
maybe_update_heartbeat_peers may remove one peer and never add in.
  if that peer is in failure_pending but that peer is not really has problem,
  failure_pending will hold that failure info until it really down or add in again.

  once ms_handle_connect called, these pending failure will be resent again, which may increase recover time after switch shutdown and then boot.

  2019-09-12 09:44:47.080933 7f1fad781700 10 osd.13 6175 ms_handle_connect on mon
  2019-09-12 09:44:47.080937 7f1fad781700 10 osd.13 6175 send_alive up_thru currently 6159 want 6155
  2019-09-12 09:44:47.080945 7f1fad781700 10 osd.13 6175 requeue_failures 0 + 1 -> 1

Signed-off-by: 宋顺10180185 <song.shun3@zte.com.cn>

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`

</details>

### CODE REVIEW

**Comment by shun-s** [comment]
2019-09-12 08:55:33.298928 7fb2907f1700  7 mon.host-192-168-9-12@0(leader).osd e6175 prepare_update osd_failure(failed timeout osd.28 [2025:3100::16]:6818/37879 for **51663sec** e6175 v6175) v3 from osd.18 [2025:3100::12]:6816/1679103
2019-09-12 08:56:03.299572 7fb2907f1700  7 mon.host-192-168-9-12@0(leader).osd e6175 prepare_update osd_failure(failed timeout osd.28 [2025:3100::16]:6818/37879 for **51693sec** e6175 v6175) v3 from osd.18 [2025:3100::12]:6816/1679103

**Comment by shun-s** [comment]
@xiexingguo  please take a review

File: src/osd/OSD.cc [hunk 1]
```diff
@@ -4505,6 +4505,16 @@ void OSD::reset_heartbeat_peers(bool all)
       it++;
     }
   }
+
+  // clean up isolated failure info
+  for (map<int,pair<utime_t,entity_inst_t> >::iterator p =
+      failure_pending.begin(); p != failure_pending.end();) {
+
+    if (heartbeat_peers.find(p->first) == heartbeat_peers.end()) {
+      failure_pending.erase(p);
```
Comment (xiexingguo) [inline src/osd/OSD.cc]: this looks the wrong way to do the erase and loop simutaneously.

e.g., 
```c++
if (heartbeat_peers.find(p->first) == heartbeat_peers.end()) {
  failure_pending.erase(p++);
} else {
  p++;
}
```

File: src/osd/OSD.cc [hunk 2]
```diff
@@ -4505,6 +4505,16 @@ void OSD::reset_heartbeat_peers(bool all)
       it++;
     }
   }
+
+  // clean up isolated failure info
```
Comment (xiexingguo) [inline src/osd/OSD.cc]: s/isolated/stale/

File: src/osd/OSD.cc [hunk 3]
```diff
@@ -4505,6 +4505,16 @@ void OSD::reset_heartbeat_peers(bool all)
...
+  for (map<int,pair<utime_t,entity_inst_t> >::iterator p =
```
Comment (xiexingguo) [inline src/osd/OSD.cc]: auto p

Comment (shun-s) [reply src/osd/OSD.cc]: done

File: src/osd/OSD.cc (hunk 2 above)
Comment (shun-s) [reply src/osd/OSD.cc]: done

File: src/osd/OSD.cc (hunk 1 above)
Comment (shun-s) [reply src/osd/OSD.cc]: done

**Comment by xiexingguo** [comment]
@liewegas I've verified this patch did fix some corner cases in our test lab. Mind double-checking this?

**Comment by liewegas** [comment]
jenkins test make check

**Comment by liewegas** [comment]
jenkins test make check

**Comment by tchaikov** [comment]
- http://pulpito.ceph.com/kchai-2019-09-25_01:32:27-rados-wip-kefu-testing-2019-09-24-1556-distro-basic-mira/
- http://pulpito.ceph.com/kchai-2019-09-27_03:53:11-rados-wip-kefu-testing-2019-09-24-1556-distro-basic-smithi/
